test/cpp/end2end/mtls_end2end_test.cc

```diff
@@ -0,0 +1,409 @@
+          std::vector<experimental::IdentityKeyCertPair> client_pair_list;
+          client_pair_list.emplace_back(client_pair);
+          channel_certificate_provider =
+              std::make_shared<experimental::StaticDataCertificateProvider>(
+                  root_certs, client_pair_list);
+          break;
+        }
+        case SecurityPrimitives::FILE_PROVIDER: {
+          channel_certificate_provider =
+              std::make_shared<FileWatcherCertificateProvider>(
+                  CLIENT_KEY_PATH, CLIENT_CERT_PATH, CA_CERT_PATH, 1);
+          break;
+        }
+      }
+      // Set up client certificate verifier.
+      std::shared_ptr<experimental::CertificateVerifier>
+          client_certificate_verifier;
+      switch (GetParam().client_verifier_type()) {
+        case SecurityPrimitives::EXTERNAL_SYNC_VERIFIER: {
+          client_certificate_verifier =
+              experimental::ExternalCertificateVerifier::Create<
+                  SyncCertificateVerifier>(true);
+          break;
+        }
+        case SecurityPrimitives::EXTERNAL_ASYNC_VERIFIER: {
+          client_certificate_verifier =
+              experimental::ExternalCertificateVerifier::Create<
+                  AsyncCertificateVerifier>(true);
+          break;
+        }
+        case SecurityPrimitives::HOSTNAME_VERIFIER: {
+          client_certificate_verifier =
+              std::make_shared<experimental::HostNameCertificateVerifier>();
+          break;
+        }
+        default: {
+          break;
+        }
+      }
+      // Configure tls credential options for each stub. Each stub connects to
+      // a separate port on the server.
+      TlsChannelCredentialsOptions channel_creds_options;
+      channel_creds_options.set_certificate_provider(
+          channel_certificate_provider);
+      channel_creds_options.watch_identity_key_cert_pairs();
+      channel_creds_options.watch_root_certs();
+      if (GetParam().client_verifier_type() !=
+          SecurityPrimitives::DEFAULT_VERIFIER) {
+        channel_creds_options.set_certificate_verifier(
+            client_certificate_verifier);
+        // When using a customized external verifier, we need to disable the
+        // per-host check.
+        if (GetParam().client_verifier_type() !=
+            SecurityPrimitives::HOSTNAME_VERIFIER) {
+          channel_creds_options.set_check_call_host(false);
+        }
+      }
+      stubs_.push_back(EchoTestService::NewStub(::grpc::CreateCustomChannel(
+          server_addresses_[i],
+          ::grpc::experimental::TlsCredentials(channel_creds_options), args)));
+    }
+  }
+
+  void TearDown() override {
+    server_->Shutdown();
+    server_thread_.join();
+  }
+
+  void RunServerLoop() { server_->Wait(); }
+
+  const std::string client_method_name_ = "grpc.testing.EchoTestService/Echo";
+  const std::string server_method_name_ = "grpc.testing.EchoTestService/Echo";
+
+  std::vector<int> ports_;
```

Comment:
Instead of having many different vectors here, please create a struct with all of the fields needed for each listening port, and then have a vector of those structs.